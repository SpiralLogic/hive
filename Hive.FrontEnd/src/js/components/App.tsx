import {GameArea} from './GameArea';
import { h } from 'preact';

export default () => {
    return <GameArea/>;
};
