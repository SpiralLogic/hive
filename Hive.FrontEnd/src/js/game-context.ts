import { createContext, useContext, useEffect, useMemo, useState } from 'preact/compat';
import { GameState, Hexagon, Move, MoveTile, Player } from './domain';
import { Engine } from './game-engine';

interface GameContext {
    players: Player[];
    hexagons: Hexagon[];
    moveTile: MoveTile;
}

export const useNewHiveContext = (): [boolean, GameContext?] => {
    const [loading, setLoading] = useState(true);
    const [gameState, setGameState] = useState<GameState>(undefined!);

    useEffect(() => {
        Engine.initialState()
            .then((state) => {
                setGameState(state);
                return setLoading(false);
            })
            .catch((reason) => {
                throw new Error(reason);
            });
    }, []);

    const moveTile = useMemo(() => (move: Move) => {
        Engine.moveTile(move)
            .then((state) => {
                return setGameState(state);
            })
            .catch((reason) => {
                throw new Error(reason);
            });
    }, [setGameState]);

    if (loading) return [loading];

    return [loading, { ...gameState, moveTile }];
};

export const HiveContext = createContext<GameContext>({
    hexagons: [],
    players: [],
    moveTile: () => undefined,
});

export const useHiveContext = () => useContext(HiveContext);
// @ts-ignore
HiveContext.displayName = 'Hive Context';
