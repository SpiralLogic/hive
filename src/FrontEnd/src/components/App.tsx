import { Fragment, FunctionComponent, h } from 'preact';
import { GameId, GameState, PlayerId } from '../domain';
import { MoveEvent, TileEvent } from '../utilities/hive-dispatcher';
import { addHiveEventListener, useHiveDispatcher } from '../utilities/hooks';
import { opponentSelectionHandler } from '../utilities/handlers';
import { useEffect, useState } from 'preact/hooks';
import Engine from '../utilities/game-engine';
import GameArea from './GameArea';
import Links from './Links';
import Rules from './Rules';

const App: FunctionComponent = () => {
  const [gameState, updateGameState] = useState<GameState | undefined>(undefined);
  const [playerId, setPlayerId] = useState<PlayerId>(0);
  const [showHelp, setShowHelp] = useState<boolean>(true);

  const initializeGame = (
    getInitial: (gameId: GameId) => Promise<GameState>,
    gameId: string,
    routePlayerId: string
  ) =>
    getInitial(gameId).then((gameState) => {
      const { gameId } = gameState;
      const [player] = gameState.players;
      const playerId = (Number(routePlayerId) as PlayerId) ? Number(routePlayerId) : player.id;
      updateGameState(gameState);
      setPlayerId(playerId);
      window.history.replaceState({ playerId, gameId }, document.title, `/game/${gameId}/${playerId}`);
    });

  useEffect(() => {
    const [, route, gameId, routePlayerId] = window.location.pathname.split('/');
    const loadExistingGame = gameId && route === 'game';
    const getInitial = loadExistingGame ? Engine.getExistingGame : Engine.getNewGame;
    initializeGame(getInitial, gameId, routePlayerId).then();
  }, []);

  if (gameState === undefined) return <h1>loading !</h1>;
  const hiveDispatcher = useHiveDispatcher();

  addHiveEventListener<MoveEvent>('move', async (event) => {
    const newGameState = await Engine.moveTile(gameState.gameId, event.move);
    updateGameState(newGameState);
  });

  useEffect(() => {
    const { closeConnection, sendSelection } = Engine.connectGame(gameState.gameId, {
      updateGameState,
      opponentSelection: opponentSelectionHandler,
    });

    const selectionChangeHandler = (event: TileEvent) => {
      if (sendSelection) {
        sendSelection('select', event.tile);
      }
    };

    const deselectionChangeHandler = (event: TileEvent) => {
      if (sendSelection) {
        sendSelection('deselect', event.tile);
      }
    };

    hiveDispatcher.add<TileEvent>('tileSelected', selectionChangeHandler);
    hiveDispatcher.add<TileEvent>('tileDeselected', deselectionChangeHandler);

    return () => {
      hiveDispatcher.remove<TileEvent>('tileSelected', selectionChangeHandler);
      hiveDispatcher.remove<TileEvent>('tileDeselected', deselectionChangeHandler);
      closeConnection().then();
    };
  }, []);

  return (
    <>
      <GameArea players={gameState.players} cells={gameState.cells} playerId={playerId} />
      <Links setShowHelp={setShowHelp} />
      <Rules showHelp={showHelp} setShowHelp={setShowHelp} />
    </>
  );
};

App.displayName = 'App';
export default App;
