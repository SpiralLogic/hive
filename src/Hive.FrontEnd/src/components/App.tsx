import '../css/app.css';

import { FunctionComponent } from 'preact';
import { useEffect, useState } from 'preact/hooks';

import { GameState } from '../domain';
import { HexEngine, HexServerConnectionFactory } from '../domain/engine';
import { addServerHandlers, opponentConnectedHandler, opponentSelectionHandler } from '../utilities/handlers';
import GameArea from './GameArea';
import { getHiveDispatcher } from '../utilities/dispatcher';

const App: FunctionComponent<{ engine: HexEngine; connectionFactory: HexServerConnectionFactory }> = (
  properties
) => {
  const { engine, connectionFactory } = properties;
  const [gameState, updateHandler] = useState<GameState | undefined>();
  const [fetchStatus, setFetchStatus] = useState('loading !');
  const hiveDispatcher = getHiveDispatcher();

  useEffect(() => {
    engine.initialGame
      .then((initialGameState) => {
        window.history.replaceState(
          { currentPlayer: engine.currentPlayer, gameId: initialGameState.gameId },
          document.title,
          `/game/${initialGameState.gameId}/${engine.currentPlayer}${document.location.search}`
        );
        updateHandler(initialGameState);
      })
      .catch((error: Error) => setFetchStatus(error.message));
  }, [engine.currentPlayer, engine.initialGame, hiveDispatcher]);

  useEffect(() => {
    if (!gameState?.gameId)
      return () => {
        /* no clean up */
      };
    const serverConnection = connectionFactory({
      currentPlayer: engine.currentPlayer,
      gameId: gameState.gameId,
      updateHandler,
      opponentSelectionHandler,
      opponentConnectedHandler,
    });
    serverConnection.connectGame().catch(() => {
      /* needs to be handled */
    });

    const removeServerHandlers = addServerHandlers(
      serverConnection.sendSelection,
      gameState.gameId,
      updateHandler,
      engine.move,
      engine.currentPlayer === 0
    );

    return (): void => {
      removeServerHandlers();
      serverConnection.closeConnection().catch(() => {
        /* needs to be handled */
      });
    };
  }, [gameState?.gameId, connectionFactory, engine.currentPlayer, engine.move]);

  if (gameState === undefined)
    return (
      <h1 class="loading">
        {fetchStatus}
        <br />
        (or broken)
      </h1>
    );

  return (
    <GameArea
      players={gameState.players}
      cells={gameState.cells}
      gameStatus={gameState.gameStatus}
      currentPlayer={engine.currentPlayer}
    />
  );
};

App.displayName = 'App';
export default App;
