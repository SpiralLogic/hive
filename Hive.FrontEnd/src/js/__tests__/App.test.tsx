import {render} from '@testing-library/preact';
import { h } from 'preact';
import {useHiveContext, useNewHiveContext} from '../game-context';
import App from '../components/App';
import Mock = jest.Mock;

jest.mock('../game-context', () => {
    const emptyCell = {coordinates: {q: 0, r: 0}, tiles: []};
    const player = {id: 2, name: 'Player 2', availableTiles: []};
    const context = {hexagons: [emptyCell], players: [player], moveTile: jest.fn()};
    return {
        useNewHiveContext: jest.fn(),
        useHiveContext: jest.fn(),
        HiveContext: context
    };
});

test('new context is created', () => {
    (useNewHiveContext as Mock).mockReturnValue([true, undefined]);
    render(<App/>);
    expect(useNewHiveContext).toHaveBeenCalled();
});

test('renders game area with context', () => {
    const emptyCell = {coordinates: {q: 0, r: 0}, tiles: []};
    const player = {id: 2, name: 'Player 2', availableTiles: []};
    const context = {hexagons: [emptyCell], players: [player], moveTile: jest.fn()};
    (useNewHiveContext as Mock).mockReturnValue([false, context]);
    (useHiveContext as Mock).mockReturnValue(context);
    render(<App/>);
    expect(document.querySelectorAll('.hive')).toHaveLength(1);
});

test('shows loading', async () => {
    (useNewHiveContext as Mock).mockReturnValue([true, undefined]);
    expect(render(<App/>).container).toMatchSnapshot();
});

test('shows error when context is missing', async () => {
    (useNewHiveContext as Mock).mockReturnValue([false, undefined]);
    expect(render(<App/>).container).toMatchSnapshot();
});

test('shows game when loaded', async () => {
    expect(render(<App/>).container).toMatchSnapshot();
});
