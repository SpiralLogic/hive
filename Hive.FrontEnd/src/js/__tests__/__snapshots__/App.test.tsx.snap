// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows error when context is missing 1`] = `
<div>
  <h1>
    Bad Times !
  </h1>
</div>
`;

exports[`shows game when loaded 1`] = `
<div>
  <h1>
    Bad Times !
  </h1>
</div>
`;

exports[`shows loading 1`] = `
<div>
  <h1>
    loading
  </h1>
</div>
`;
